oracle.go

```diff
@@ -0,0 +1,336 @@
+	}
+	foundIdHex := make(map[string]bool, len(objectIds))
+	for iter.Next(&txnDoc) {
+		foundIdHex[txnDoc.Id.Hex()] = true
+	}
+	if err := iter.Close(); err != nil {
+		if err != mgo.ErrNotFound {
+			// Not found is ok, the transactions may not be complete
+			return nil, err
+		}
+	}
+	result := make(map[string]bool, len(foundIdHex))
+	// because multiple tokens could map to a single txn, we iterate the
+	// passed in tokens instead of caching them in the map.
+	for _, token := range tokens {
+		objIdHex := txnTokenToId(token).Hex()
+		if foundIdHex[objIdHex] {
+			result[token] = true
+		}
+	}
+	o.checkedTokens += uint64(len(tokens))
+	o.completedTokens += uint64(len(result))
+	o.foundTxns += uint64(len(foundIdHex))
+	return result, nil
+}
+
+// RemoveTxns can be used to flag that a given transaction should not
+// be considered part of the valid set.
+func (o *DBOracle) RemoveTxns(txnIds []bson.ObjectId) error {
+	_, err := o.working.RemoveAll(bson.M{"_id": bson.M{"$in": txnIds}})
+	if err != nil {
+		return fmt.Errorf("error removing transaction ids: %v", err)
+	}
+	return nil
+}
+
+type dbIterWrapper struct {
+	iter *mgo.Iter
+}
+
+var _ OracleIterator = (*dbIterWrapper)(nil)
+
+var EOF = fmt.Errorf("end of transaction ids")
+
+func (d *dbIterWrapper) Next() (bson.ObjectId, error) {
+	var txnId struct {
+		Id bson.ObjectId `bson:"_id"`
+	}
+	if d.iter.Next(&txnId) {
+		return txnId.Id, nil
+	}
+	if err := d.iter.Close(); err != nil {
+		return txnId.Id, err
+	}
+	return txnId.Id, EOF
+}
+func (d *dbIterWrapper) Close() error {
+	return d.iter.Close()
+}
+
+// IterTxns lets you iterate over all of the transactions that have
+// not been removed.
+func (o *DBOracle) IterTxns() (OracleIterator, error) {
+	iter := o.working.Find(nil).Select(bson.M{"_id": 1}).Iter()
+	return &dbIterWrapper{iter: iter}, nil
+}
+
+// MemOracle uses a temporary table on disk to track what transactions are
+// considered completed and purgeable.
+type MemOracle struct {
+	txns            *mgo.Collection
+	completed       map[bson.ObjectId]bool
+	checkedTokens   uint64
+	completedTokens uint64
+	foundTxns       uint64
+}
+
+// NewMemOracle uses an in-memory map to manage the queue of  remaining
+// transactions.
+func NewMemOracle(txns *mgo.Collection) *MemOracle {
+	return &MemOracle{
+		txns: txns,
+	}
+}
+
+var _ Oracle = (*MemOracle)(nil)
+
+func (o *MemOracle) Prepare() (func(), error) {
+	if o.completed != nil {
+		return noopCleanup, fmt.Errorf("Prepare called twice")
+	}
+	// Load the ids of all completed and aborted txns into a separate
+	// temporary collection.
+	logger.Debugf("loading all completed transactions")
+	pipe := o.txns.Pipe([]bson.M{
+		// This used to use $in but that's much slower than $gte.
+		{"$match": bson.M{"s": bson.M{"$gte": taborted}}},
+		{"$project": bson.M{"_id": 1}},
+	})
+	pipe.Batch(maxBatchDocs)
+	pipe.AllowDiskUse()
+	var txnId struct {
+		Id bson.ObjectId `bson:"_id"`
+	}
+	completed := make(map[bson.ObjectId]bool, 1000)
```

Comment:
I had run tests on large-ish maps when I wrote this. It was more efficient, but only about 10% IIRC. I can certainly use it if you prefer.
```
$ go run memtest.go
alloc: 284861288, 278184.9k, 271.7M
vs
$ go run memtest.go
alloc: 315390408, 307998.4k, 300.8M
```

For this code:
```
package main

import (
        "fmt"
        "runtime"
)

func main() {
        // val := struct {}{}
        // content := make(map[uint64]struct{})
        val := true
        content := make(map[uint64]bool)
        for i := uint64(0); i < 10000000; i++ {
                content[i] = val
        }
        var stats runtime.MemStats
        runtime.ReadMemStats(&stats)
        fmt.Printf("alloc: %d, %.1fk, %.1fM\n", stats.Alloc, float64(stats.Alloc)/1024., float64(stats.Alloc)/(1024*1024))
}
```

You're right that 10% isn't nothing, so I'll go ahead and use it. I just wanted you to be aware that it wasn't a very large difference. (I imagine it depends heavily on the size of your keys vs the size of a bool, but all of the map overhead, etc, dominate the size of the map, vs the size of the bool vs struct{} on the output.)